File: src/graph_theory/flow.py
"""
Maximum flow and minimum cut.

Lester Ford and Delbert Fulkerson, RAND Corporation, 1956. Their paper
"Maximal Flow Through a Network" (Canadian Journal of Mathematics, 1956)
proved the max-flow / min-cut theorem and gave the augmenting-path
algorithm. They were working in the same building at RAND as Ted Harris
and Frank Ross, who had just produced a classified study modelling the
Soviet-and-Eastern-European rail network as a graph and asking which
links, if cut, would sever supply from the Soviet interior to its
satellites. The theorem answered their question.

This module implements Edmonds-Karp (Ford-Fulkerson with BFS for the
augmenting path). BFS guarantees termination on irrational capacities
and gives O(V * E^2) complexity. Depth-first augmenting paths can fail
to terminate on real-valued capacities; BFS is the standard pedagogical
and practical choice.

The min-cut is recovered by running max-flow, then doing a single BFS
from the source on the residual graph. Nodes reachable in the residual
form one side of the cut; the cut edges are the original edges that
cross from reachable to unreachable.

Usage::

    from src.graph_theory.flow import FlowNetwork, max_flow, min_cut

    net = FlowNetwork()
    net.add_edge("s", "a", 10)
    net.add_edge("a", "t", 10)
    value, flow = max_flow(net, "s", "t")
    cut_value, cut_edges = min_cut(net, "s", "t")
    assert value == cut_value
"""

from collections import defaultdict, deque
from dataclasses import dataclass, field


@dataclass
class FlowStep:
    """One iteration of Edmonds-Karp. See ``max_flow_steps``."""

    path: list[str]
    bottleneck: float
    flow_value: float
    residual: dict[str, dict[str, float]]


@dataclass
class FlowNetwork:
    """
    Directed graph with non-negative edge capacities.

    Edges are directed and asymmetric: ``add_edge(u, v, c)`` adds a
    single edge from ``u`` to ``v`` with capacity ``c``. The reverse
    edge used by the algorithm lives in the residual graph, not here.

    Parallel edges in the same direction are merged by summing their
    capacities; this matches how flow networks are usually drawn in
    practice and avoids pathological cases for the residual construction.

    Attributes:
        _cap: Capacity mapping ``_cap[u][v]`` is the capacity of the
              edge from ``u`` to ``v``. Prefer the public methods.
    """

    _cap: dict[str, dict[str, float]] = field(
        default_factory=lambda: defaultdict(dict)
    )

    def add_node(self, node: str) -> None:
        """Add a node with no edges. Idempotent."""
        if node not in self._cap:
            self._cap[node] = {}

    def add_edge(self, u: str, v: str, capacity: float) -> None:
        """
        Add a directed edge ``u -> v`` with the given capacity.

        If an edge from ``u`` to ``v`` already exists, the capacities
        are summed.

        Raises:
            ValueError: if ``capacity`` is negative, or if ``u == v``
                (self-loops carry no flow and usually indicate a bug).
        """
        if capacity < 0:
            raise ValueError(
                f"Edge ({u!r}, {v!r}) has negative capacity {capacity}; "
                "flow networks require non-negative capacities."
            )
        if u == v:
            raise ValueError(
                f"Self-loop on {u!r} is not a valid flow edge."
            )
        self.add_node(u)
        self.add_node(v)
        self._cap[u][v] = self._cap[u].get(v, 0.0) + capacity

    def capacity(self, u: str, v: str) -> float:
        """Capacity of edge ``u -> v``; 0 if no such edge exists."""
        return self._cap.get(u, {}).get(v, 0.0)

    @property
    def nodes(self) -> list[str]:
        """All nodes in the network."""
        return list(self._cap.keys())

    @property
    def edges(self) -> list[tuple[str, str, float]]:
        """All directed edges as ``(u, v, capacity)`` triples."""
        out: list[tuple[str, str, float]] = []
        for u, vs in self._cap.items():
            for v, c in vs.items():
                out.append((u, v, c))
        return out


def _build_residual(net: FlowNetwork) -> dict[str, dict[str, float]]:
    """
    Initialise a residual graph from a flow network.

    The residual graph has the same forward capacities as the original
    plus a reverse edge for every forward edge, initially at zero. As
    flow is pushed, forward capacity decreases and reverse capacity
    grows, allowing the algorithm to "cancel" flow along an earlier
    path if a better one is found later.
    """
    residual: dict[str, dict[str, float]] = defaultdict(dict)
    for u in net.nodes:
        residual[u] = {}
    for u, v, c in net.edges:
        residual[u][v] = residual[u].get(v, 0.0) + c
        if u not in residual[v]:
            residual[v][u] = 0.0
    return residual


def _bfs_augmenting_path(
    residual: dict[str, dict[str, float]], source: str, sink: str
) -> list[str] | None:
    """
    Return the shortest (in edge count) augmenting path, or None.

    BFS on the residual graph, considering only edges with positive
    residual capacity. The returned path starts at ``source`` and ends
    at ``sink``.
    """
    predecessors: dict[str, str | None] = {source: None}
    queue: deque[str] = deque([source])
    while queue:
        u = queue.popleft()
        if u == sink:
            break
        for v, cap in residual[u].items():
            if cap > 0 and v not in predecessors:
                predecessors[v] = u
                queue.append(v)
    if sink not in predecessors:
        return None
    path: list[str] = []
    node: str | None = sink
    while node is not None:
        path.append(node)
        node = predecessors[node]
    path.reverse()
    return path


def max_flow(
    net: FlowNetwork, source: str, sink: str
) -> tuple[float, dict[tuple[str, str], float]]:
    """
    Maximum flow from ``source`` to ``sink`` using Edmonds-Karp.

    Args:
        net: The flow network. Capacities must be non-negative.
        source: The node flow originates from.
        sink: The node flow terminates at.

    Returns:
        A tuple ``(value, flow)``:
          - ``value`` is the maximum flow value.
          - ``flow[(u, v)]`` is the flow along the original edge
            ``u -> v``. Only original edges appear; reverse edges are
            an implementation detail of the residual graph.

    Raises:
        ValueError: if ``source`` or ``sink`` is not in ``net``, or if
            ``source == sink``.
    """
    if source not in net.nodes:
        raise ValueError(f"Source {source!r} not in network.")
    if sink not in net.nodes:
        raise ValueError(f"Sink {sink!r} not in network.")
    if source == sink:
        raise ValueError("Source and sink must be distinct.")

    residual = _build_residual(net)
    original_caps = {(u, v): c for u, v, c in net.edges}
    value = 0.0

    while True:
        path = _bfs_augmenting_path(residual, source, sink)
        if path is None:
            break
        bottleneck = min(
            residual[path[i]][path[i + 1]] for i in range(len(path) - 1)
        )
        for i in range(len(path) - 1):
            u, v = path[i], path[i + 1]
            residual[u][v] -= bottleneck
            residual[v][u] += bottleneck
        value += bottleneck

    flow: dict[tuple[str, str], float] = {}
    for (u, v), cap in original_caps.items():
        flow[(u, v)] = cap - residual[u][v]
    return value, flow


def min_cut(
    net: FlowNetwork, source: str, sink: str
) -> tuple[float, set[tuple[str, str]]]:
    """
    Minimum s-t cut value and the edges crossing the cut.

    By the max-flow / min-cut theorem (Ford and Fulkerson, 1956) the
    value returned here equals ``max_flow(net, source, sink)[0]``.

    The cut is recovered by running max-flow, then doing a single BFS
    from ``source`` on the final residual graph. Nodes reachable in
    the residual form the source side; the cut edges are the *original*
    edges that cross from reachable to unreachable.

    Args:
        net: The flow network.
        source: The source node.
        sink: The sink node.

    Returns:
        A tuple ``(value, edges)`` where ``edges`` is the set of
        directed original edges ``(u, v)`` with ``u`` reachable and
        ``v`` unreachable in the residual.
    """
    value, _ = max_flow(net, source, sink)

    residual = _build_residual(net)
    # Replay the augmenting paths to get the final residual. Cheaper
    # than threading the residual out of max_flow and keeps that
    # function's contract narrow.
    original_caps = {(u, v): c for u, v, c in net.edges}
    while True:
        path = _bfs_augmenting_path(residual, source, sink)
        if path is None:
            break
        bottleneck = min(
            residual[path[i]][path[i + 1]] for i in range(len(path) - 1)
        )
        for i in range(len(path) - 1):
            u, v = path[i], path[i + 1]
            residual[u][v] -= bottleneck
            residual[v][u] += bottleneck

    reachable: set[str] = set()
    queue: deque[str] = deque([source])
    reachable.add(source)
    while queue:
        u = queue.popleft()
        for v, cap in residual[u].items():
            if cap > 0 and v not in reachable:
                reachable.add(v)
                queue.append(v)

    cut_edges: set[tuple[str, str]] = set()
    for (u, v) in original_caps:
        if u in reachable and v not in reachable:
            cut_edges.add((u, v))
    return value, cut_edges


def max_flow_steps(
    net: FlowNetwork, source: str, sink: str
) -> list[FlowStep]:
    """
    Record each augmenting-path iteration of Edmonds-Karp.

    Useful for visualisation. Each step captures the augmenting path,
    the bottleneck capacity along it, the cumulative flow value after
    that augmentation, and a snapshot of the residual graph.

    The returned list always has at least one entry: the initial state
    (empty path, zero flow, initial residual) at index 0.
    """
    if source not in net.nodes:
        raise ValueError(f"Source {source!r} not in network.")
    if sink not in net.nodes:
        raise ValueError(f"Sink {sink!r} not in network.")
    if source == sink:
        raise ValueError("Source and sink must be distinct.")

    residual = _build_residual(net)
    steps: list[FlowStep] = [
        FlowStep(
            path=[],
            bottleneck=0.0,
            flow_value=0.0,
            residual={u: dict(vs) for u, vs in residual.items()},
        )
    ]
    value = 0.0
    while True:
        path = _bfs_augmenting_path(residual, source, sink)
        if path is None:
            break
        bottleneck = min(
            residual[path[i]][path[i + 1]] for i in range(len(path) - 1)
        )
        for i in range(len(path) - 1):
            u, v = path[i], path[i + 1]
            residual[u][v] -= bottleneck
            residual[v][u] += bottleneck
        value += bottleneck
        steps.append(
            FlowStep(
                path=list(path),
                bottleneck=bottleneck,
                flow_value=value,
                residual={u: dict(vs) for u, vs in residual.items()},
            )
        )
    return steps

File: website/src/components/interactive/graph-theory/flow.ts
// Flow networks, max-flow, min-cut for the Graph Theory / Soviet Rail lesson.
// Mirrors src/graph_theory/flow.py. Keep the two in sync when changing semantics.

export interface FlowEdge {
  u: string;
  v: string;
  capacity: number;
}

export class FlowNetwork {
  private cap = new Map<string, Map<string, number>>();

  addNode(n: string) {
    if (!this.cap.has(n)) this.cap.set(n, new Map());
  }

  addEdge(u: string, v: string, capacity: number) {
    if (capacity < 0) {
      throw new Error(
        `Edge (${u}, ${v}) has negative capacity ${capacity}; flow networks require non-negative capacities.`
      );
    }
    if (u === v) {
      throw new Error(`Self-loop on ${u} is not a valid flow edge.`);
    }
    this.addNode(u);
    this.addNode(v);
    const existing = this.cap.get(u)!.get(v) ?? 0;
    this.cap.get(u)!.set(v, existing + capacity);
  }

  capacity(u: string, v: string): number {
    return this.cap.get(u)?.get(v) ?? 0;
  }

  nodes(): string[] {
    return [...this.cap.keys()];
  }

  edges(): FlowEdge[] {
    const out: FlowEdge[] = [];
    for (const [u, vs] of this.cap) {
      for (const [v, c] of vs) out.push({ u, v, capacity: c });
    }
    return out;
  }
}

type Residual = Map<string, Map<string, number>>;

function buildResidual(net: FlowNetwork): Residual {
  const r: Residual = new Map();
  for (const u of net.nodes()) r.set(u, new Map());
  for (const { u, v, capacity } of net.edges()) {
    const row = r.get(u)!;
    row.set(v, (row.get(v) ?? 0) + capacity);
    const rev = r.get(v)!;
    if (!rev.has(u)) rev.set(u, 0);
  }
  return r;
}

function bfsAugmentingPath(
  residual: Residual,
  source: string,
  sink: string
): string[] | null {
  const predecessors = new Map<string, string | null>();
  predecessors.set(source, null);
  const queue: string[] = [source];
  while (queue.length > 0) {
    const u = queue.shift()!;
    if (u === sink) break;
    for (const [v, cap] of residual.get(u)!) {
      if (cap > 0 && !predecessors.has(v)) {
        predecessors.set(v, u);
        queue.push(v);
      }
    }
  }
  if (!predecessors.has(sink)) return null;
  const path: string[] = [];
  let node: string | null = sink;
  while (node !== null) {
    path.push(node);
    node = predecessors.get(node)!;
  }
  path.reverse();
  return path;
}

function edgeKey(u: string, v: string): string {
  return `${u}->${v}`;
}

export interface MaxFlowResult {
  value: number;
  /** Flow on each original edge, keyed by "u->v". */
  flow: Map<string, number>;
}

export function maxFlow(
  net: FlowNetwork,
  source: string,
  sink: string
): MaxFlowResult {
  if (!net.nodes().includes(source)) {
    throw new Error(`Source ${source} not in network.`);
  }
  if (!net.nodes().includes(sink)) {
    throw new Error(`Sink ${sink} not in network.`);
  }
  if (source === sink) {
    throw new Error("Source and sink must be distinct.");
  }

  const residual = buildResidual(net);
  const originalCaps = new Map<string, number>();
  for (const { u, v, capacity } of net.edges()) {
    originalCaps.set(edgeKey(u, v), capacity);
  }

  let value = 0;
  while (true) {
    const path = bfsAugmentingPath(residual, source, sink);
    if (path === null) break;
    let bottleneck = Infinity;
    for (let i = 0; i + 1 < path.length; i++) {
      const c = residual.get(path[i])!.get(path[i + 1])!;
      if (c < bottleneck) bottleneck = c;
    }
    for (let i = 0; i + 1 < path.length; i++) {
      const u = path[i];
      const v = path[i + 1];
      residual.get(u)!.set(v, residual.get(u)!.get(v)! - bottleneck);
      residual.get(v)!.set(u, (residual.get(v)!.get(u) ?? 0) + bottleneck);
    }
    value += bottleneck;
  }

  const flow = new Map<string, number>();
  for (const [key, cap] of originalCaps) {
    const [u, v] = key.split("->");
    const remaining = residual.get(u)!.get(v) ?? 0;
    flow.set(key, cap - remaining);
  }
  return { value, flow };
}

export interface MinCutResult {
  value: number;
  /** Directed original edges crossing the cut (source side -> sink side). */
  edges: Set<string>;
  /** Nodes on the source side of the cut. */
  sourceSide: Set<string>;
}

export function minCut(
  net: FlowNetwork,
  source: string,
  sink: string
): MinCutResult {
  const { value } = maxFlow(net, source, sink);

  // Replay the augmentations to recover the final residual.
  const residual = buildResidual(net);
  const originalCaps = new Map<string, number>();
  for (const { u, v, capacity } of net.edges()) {
    originalCaps.set(edgeKey(u, v), capacity);
  }
  while (true) {
    const path = bfsAugmentingPath(residual, source, sink);
    if (path === null) break;
    let bottleneck = Infinity;
    for (let i = 0; i + 1 < path.length; i++) {
      const c = residual.get(path[i])!.get(path[i + 1])!;
      if (c < bottleneck) bottleneck = c;
    }
    for (let i = 0; i + 1 < path.length; i++) {
      const u = path[i];
      const v = path[i + 1];
      residual.get(u)!.set(v, residual.get(u)!.get(v)! - bottleneck);
      residual.get(v)!.set(u, (residual.get(v)!.get(u) ?? 0) + bottleneck);
    }
  }

  const reachable = new Set<string>([source]);
  const queue: string[] = [source];
  while (queue.length > 0) {
    const u = queue.shift()!;
    for (const [v, cap] of residual.get(u)!) {
      if (cap > 0 && !reachable.has(v)) {
        reachable.add(v);
        queue.push(v);
      }
    }
  }

  const edges = new Set<string>();
  for (const key of originalCaps.keys()) {
    const [u, v] = key.split("->");
    if (reachable.has(u) && !reachable.has(v)) edges.add(key);
  }
  return { value, edges, sourceSide: reachable };
}

// ── Step-by-step variant for the visualiser ─────────────────────────

export interface FlowStep {
  /** Augmenting path found at this step (empty on the initial step). */
  path: string[];
  /** Bottleneck capacity along that path (0 on the initial step). */
  bottleneck: number;
  /** Cumulative flow value after this step. */
  flowValue: number;
  /** Residual capacity per edge after this step, keyed by "u->v" (includes reverse edges). */
  residual: Map<string, number>;
  /** Flow on each original edge after this step, keyed by "u->v". */
  flow: Map<string, number>;
}

export function maxFlowSteps(
  net: FlowNetwork,
  source: string,
  sink: string
): FlowStep[] {
  if (!net.nodes().includes(source)) {
    throw new Error(`Source ${source} not in network.`);
  }
  if (!net.nodes().includes(sink)) {
    throw new Error(`Sink ${sink} not in network.`);
  }
  if (source === sink) {
    throw new Error("Source and sink must be distinct.");
  }

  const residual = buildResidual(net);
  const originalCaps = new Map<string, number>();
  for (const { u, v, capacity } of net.edges()) {
    originalCaps.set(edgeKey(u, v), capacity);
  }

  const snapshot = (): { residual: Map<string, number>; flow: Map<string, number> } => {
    const rSnap = new Map<string, number>();
    for (const [u, vs] of residual) {
      for (const [v, c] of vs) rSnap.set(edgeKey(u, v), c);
    }
    const fSnap = new Map<string, number>();
    for (const [key, cap] of originalCaps) {
      const [u, v] = key.split("->");
      fSnap.set(key, cap - (residual.get(u)!.get(v) ?? 0));
    }
    return { residual: rSnap, flow: fSnap };
  };

  const steps: FlowStep[] = [];
  {
    const { residual: r, flow: f } = snapshot();
    steps.push({ path: [], bottleneck: 0, flowValue: 0, residual: r, flow: f });
  }

  let value = 0;
  while (true) {
    const path = bfsAugmentingPath(residual, source, sink);
    if (path === null) break;
    let bottleneck = Infinity;
    for (let i = 0; i + 1 < path.length; i++) {
      const c = residual.get(path[i])!.get(path[i + 1])!;
      if (c < bottleneck) bottleneck = c;
    }
    for (let i = 0; i + 1 < path.length; i++) {
      const u = path[i];
      const v = path[i + 1];
      residual.get(u)!.set(v, residual.get(u)!.get(v)! - bottleneck);
      residual.get(v)!.set(u, (residual.get(v)!.get(u) ?? 0) + bottleneck);
    }
    value += bottleneck;
    const { residual: r, flow: f } = snapshot();
    steps.push({ path, bottleneck, flowValue: value, residual: r, flow: f });
  }

  return steps;
}

// ── Soviet rail fixture used by the interactive demos ──────────────

export interface CityNode {
  id: string;
  name: string;
  x: number;
  y: number;
}

// Stylised layout. Moscow (source) is on the right; the Western Frontier
// sink is on the left. Positions are chosen for visual clarity, not
// geographical accuracy. The node set is loosely modelled on the 1955
// Harris-Ross study of the Soviet rail network (their study had ~44
// vertices; we use ten so the visual stays legible).
export const SOVIET_CITIES: CityNode[] = [
  { id: "MOW", name: "Moscow",    x: 560, y: 230 },
  { id: "LED", name: "Leningrad", x: 500, y:  85 },
  { id: "MSK", name: "Minsk",     x: 410, y: 190 },
  { id: "KIE", name: "Kiev",      x: 400, y: 320 },
  { id: "WAW", name: "Warsaw",    x: 300, y: 210 },
  { id: "BER", name: "Berlin",    x: 185, y: 130 },
  { id: "PRG", name: "Prague",    x: 225, y: 275 },
  { id: "BUD", name: "Budapest",  x: 360, y: 405 },
  { id: "VIE", name: "Vienna",    x: 270, y: 375 },
  { id: "FRO", name: "Frontier",  x:  75, y: 230 },
];

export interface RailLine {
  u: string;
  v: string;
  /** Default capacity, in stylised supply units per day. */
  capacity: number;
}

// Directed edges: supply flows from Moscow westward to the frontier.
// Capacities are illustrative; the Harris-Ross study used track count,
// grade, and locomotive availability to derive theirs.
export const SOVIET_LINES: RailLine[] = [
  { u: "MOW", v: "LED", capacity: 30 },
  { u: "MOW", v: "MSK", capacity: 40 },
  { u: "MOW", v: "KIE", capacity: 35 },
  { u: "LED", v: "MSK", capacity: 20 },
  { u: "LED", v: "BER", capacity: 25 },
  { u: "MSK", v: "WAW", capacity: 30 },
  { u: "KIE", v: "WAW", capacity: 20 },
  { u: "KIE", v: "BUD", capacity: 25 },
  { u: "WAW", v: "BER", capacity: 35 },
  { u: "WAW", v: "PRG", capacity: 25 },
  { u: "PRG", v: "VIE", capacity: 20 },
  { u: "PRG", v: "FRO", capacity: 30 },
  { u: "BER", v: "FRO", capacity: 40 },
  { u: "VIE", v: "FRO", capacity: 25 },
  { u: "BUD", v: "VIE", capacity: 20 },
];

export const SOVIET_SOURCE = "MOW";
export const SOVIET_SINK = "FRO";

export function buildSovietNetwork(
  capacities: Map<string, number>
): FlowNetwork {
  const net = new FlowNetwork();
  for (const c of SOVIET_CITIES) net.addNode(c.id);
  for (const line of SOVIET_LINES) {
    const key = edgeKey(line.u, line.v);
    const cap = capacities.get(key) ?? line.capacity;
    net.addEdge(line.u, line.v, cap);
  }
  return net;
}

export { edgeKey as flowEdgeKey };
